Algorithm.Framework/Selection/OpenInterestFutureUniverseSelectionModel.cs

```diff
@@ -0,0 +1,117 @@
+
+namespace QuantConnect.Algorithm.Framework.Selection
+{
+    /// <summary>
+    ///     Selects contracts in a futures universe, sorted by open interest.  This allows the selection to identifiy current
+    ///     active contract.
+    /// </summary>
+    public class OpenInterestFutureUniverseSelectionModel : FutureUniverseSelectionModel
+    {
+        private readonly int? _chainContractsLookupLimit;
+        private readonly IAlgorithm _algorithm;
+        private readonly int? _resultsLimit;
+        private readonly MarketHoursDatabase _marketHoursDatabase;
+
+        /// <summary>
+        ///     Creates a new instance of <see cref="OpenInterestFutureUniverseSelectionModel" />
+        /// </summary>
+        /// <param name="algorithm">Algorithm</param>
+        /// <param name="futureChainSymbolSelector">Selects symbols from the provided future chain</param>
+        /// <param name="chainContractsLookupLimit">Limit on how many contracts to query for open interest</param>
+        /// <param name="resultsLimit">Limit on how many contracts will be part of the universe</param>
+        public OpenInterestFutureUniverseSelectionModel(IAlgorithm algorithm, Func<DateTime, IEnumerable<Symbol>> futureChainSymbolSelector, int? chainContractsLookupLimit = 6,
+            int? resultsLimit = 1) : base(TimeSpan.FromDays(1), futureChainSymbolSelector)
+        {
+            _marketHoursDatabase = MarketHoursDatabase.FromDataFolder();
+            if (algorithm == null)
+            {
+                throw new ArgumentNullException(nameof(algorithm));
+            }
+
+            _algorithm = algorithm;
+            _resultsLimit = resultsLimit;
+            _chainContractsLookupLimit = chainContractsLookupLimit;
+        }
+
+        /// <summary>
+        ///     Defines the future chain universe filter
+        /// </summary>
+        protected override FutureFilterUniverse Filter(FutureFilterUniverse filter)
+        {
+            return filter.Contracts(FilterByOpenInterest(filter.ToDictionary(x => x, x => _marketHoursDatabase.GetEntry(x.ID.Market, x, x.ID.SecurityType).ExchangeHours)));
+        }
+
+        /// <summary>
+        ///     Filters a set of contracts based on open interest.
+        /// </summary>
+        /// <param name="contracts">Contracts to filter</param>
+        /// <returns>Filtered set</returns>
+        public IEnumerable<Symbol> FilterByOpenInterest(IReadOnlyDictionary<Symbol, SecurityExchangeHours> contracts)
+        {
+            var filtered = _chainContractsLookupLimit.HasValue ? contracts.Keys.OrderBy(x => x.ID.Date).Take(_chainContractsLookupLimit.Value) : contracts.Keys;
+            var openInterest = filtered.GroupBy(x => contracts[x]).SelectMany(g => GetOpenInterest(g.Key, g.Select(i => i))).ToDictionary(x => x.Key, x => x.Value);
+
+            decimal o;
+            filtered = filtered.OrderByDescending(s => openInterest.TryGetValue(s, out o) ? o : -1).ThenBy(x => x.ID.Date);
```

Comment:
Maybe should orderBy directly on the resulting `openInterest`? else if the symbol is not in `openInterest` it will use -1 and it might return that symbol.
What do you think about adding something like:
```
if (openInterest.Count == 0)
{
    _algorithm.Error("OpenInterestFutureUniverseSelectionModel.FilterByOpenInterest(): failed to get historical open interest, will select no symbol");
    return Enumerable.Empty<Symbol>();
}
```

Reply:
Okay, I've adopted a form of this.  Let me know if you're okay with that, and resolve if so.